src/renderer/base/renderer.cpp

```diff
@@ -78,8 +78,17 @@ Renderer::~Renderer()
             {
                 if (--tries == 0)
                 {
-                    FAIL_FAST_HR_MSG(E_UNEXPECTED, "A rendering engine required too many retries.");
+                    // Stop trying.
+                    _pThread->DisablePainting();
+                    if (_pfnRendererEnteredErrorState)
+                    {
+                        _pfnRendererEnteredErrorState();
+                    }
+                    return S_FALSE;

```

Comment:
If there's not a callback, and we `S_FALSE` here, there's effectively no way for the hosting app to recover, right? they won't even _know_ that something went wrong, right?

Should we instead still fail-fast if there's no callback?

Reply:
This is correct. The WPF control won't hook it up, for now. If we failfast here, without a callback, they'll be in the same boat as before (a failfast.)

Zoey suggested that us blacking out is fine, and the user can close/reopen the terminal if this happens.  It should be rarer with the backoff time.

Reply:
Okay if Zoey said this is okay then I'm fine with this. I figured leaving the failfast would encourage other apps (WPF) to implement a similar retry button, but if this is what they want the :shipit:

Reply:
Commenting that.